parser-typechecker/src/Unison/Names3.hs

```diff
@@ -33,6 +37,22 @@ data ResolutionFailure v a
 
 type ResolutionResult v a r = Either (Seq (ResolutionFailure v a)) r
 
+-- For all names in `ns`, (ex: foo.bar.baz), generate the list of suffixes 
+-- of that name [[foo.bar.baz], [bar.baz], [baz]]. Insert these suffixes
+-- into a multimap map along with their corresponding refs. Any suffix
+-- which is unique is added as an entry to `ns`.
+suffixify0 :: Names0 -> Names0
+suffixify0 ns = ns <> suffixNs 
+  where
+  suffixNs = names0 (R.fromList uniqueTerms) (R.fromList uniqueTypes) 
+  terms = List.multimap [ (n,ref) | (n0,ref) <- R.toList (terms0 ns), n <- Name.suffixes n0 ]
+  types = List.multimap [ (n,ref) | (n0,ref) <- R.toList (types0 ns), n <- Name.suffixes n0 ]
+  uniqueTerms = [ (n,ref) | (n, nubOrd -> [ref]) <- Map.toList terms ]
+  uniqueTypes = [ (n,ref) | (n, nubOrd -> [ref]) <- Map.toList types ]
```

Comment:
This is the main interesting function responsible for adding new name aliases for unique  suffixes. Thought: we may want to move to representing `Names` as a suffix trie or some such if we want this kind of operation to be more efficient. NBD for now though.